File: ts-client/src/curve/constant-product.ts
import sqrt from "bn-sqrt";
import BN from "bn.js";
import { SwapCurve } from ".";

// Typescript implementation of https://github.com/solana-labs/solana-program-library/blob/master/libraries/math/src/checked_ceil_div.rs#L29
function ceilDiv(lhs: BN, rhs: BN) {
  let quotient = lhs.div(rhs);
  // Avoid dividing a small number by a big one and returning 1, and instead
  // fail.
  if (quotient.eq(new BN(0))) {
    throw new Error("ceilDiv result in zero");
  }

  let remainder = lhs.mod(rhs);

  if (remainder.gt(new BN(0))) {
    quotient = quotient.add(new BN(1));
    rhs = lhs.div(quotient);
    remainder = lhs.mod(quotient);
    if (remainder.gt(new BN(0))) {
      rhs = rhs.add(new BN(1));
    }
  }

  return [quotient, rhs];
}

export class ConstantProductSwap implements SwapCurve {
  constructor() {}
  // Typescript implementation of https://github.com/solana-labs/solana-program-library/blob/master/token-swap/program/src/curve/constant_product.rs#L27
  computeOutAmount(
    sourceAmount: BN,
    swapSourceAmount: BN,
    swapDestinationAmount: BN
  ): BN {
    let invariant = swapSourceAmount.mul(swapDestinationAmount);
    let [newSwapDestinationAmount, newSwapSourceAmount] = ceilDiv(
      invariant,
      swapSourceAmount.add(sourceAmount)
    );
    let _sourceAmountSwapped = newSwapSourceAmount.sub(sourceAmount);
    let destinationAmountSwapped = swapDestinationAmount.sub(
      newSwapDestinationAmount
    );
    if (destinationAmountSwapped.eq(new BN(0))) {
      throw new Error("Swap result in zero");
    }
    return destinationAmountSwapped;
  }
  computeD(tokenAAmount: BN, tokenBAmount: BN): BN {
    return sqrt(tokenAAmount.mul(tokenBAmount));
  }
}

File: programs/amm/src/math/constant_product.rs
use super::*;
use spl_token_swap::curve::{calculator::SwapWithoutFeesResult, constant_product::swap};

pub struct ConstantProduct {}

impl SwapCurve for ConstantProduct {
    fn swap(
        &self,
        source_amount: u64,
        swap_source_amount: u64,
        swap_destination_amount: u64,
        _trade_direction: TradeDirection,
    ) -> Option<SwapResult> {
        let source_amount: u128 = source_amount.into();
        let swap_source_amount: u128 = swap_source_amount.into();
        let swap_destination_amount: u128 = swap_destination_amount.into();

        let SwapWithoutFeesResult {
            source_amount_swapped,
            destination_amount_swapped,
        } = swap(source_amount, swap_source_amount, swap_destination_amount)?;

        Some(SwapResult {
            new_swap_source_amount: swap_source_amount.checked_add(source_amount_swapped)?,
            new_swap_destination_amount: swap_destination_amount
                .checked_sub(destination_amount_swapped)?,
            source_amount_swapped,
            destination_amount_swapped,
        })
    }
}
